pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java

```diff
@@ -264,4 +267,31 @@ static boolean isFitForDictionaryBasedPlan(QueryContext queryContext, IndexSegme
     }
     return true;
   }
+
+  /**
+   * Returns dictionary based distinct plan node iff supported, else distinct plan node
+   */
+  private PlanNode getDistinctExecutionNode(IndexSegment indexSegment, QueryContext queryContext) {
+    AggregationFunction[] aggregationFunctions = queryContext.getAggregationFunctions();
+
+    // If we have gotten here, it must have been verified that there is only on aggregation function in the context
+    // and it is a DistinctAggregationFunction
+
+    DistinctAggregationFunction distinctAggregationFunction = (DistinctAggregationFunction) aggregationFunctions[0];
+    List<ExpressionContext> expressions = distinctAggregationFunction.getInputExpressions();
+
+    if (expressions.size() == 1 && queryContext.getFilter() == null) {
+      ExpressionContext expression = expressions.get(0);
+
+      if (expression.getType() == ExpressionContext.Type.IDENTIFIER) {
+        String column = expression.getIdentifier();
+        Dictionary dictionary = indexSegment.getDataSource(column).getDictionary();
+        if (dictionary != null && dictionary.isSorted()) {
```

Comment:
We should still be able to use dictionary based operator when dictionary is not sorted: create a `DistinctTable` and add all dictionary values into the table

Reply:
How would we apply limits on it?